File: Persistence/Source/Repositories/MeasurementsRepository.vb
Imports Wells.Model

Public Class MeasurementsRepository
    Inherits Repository(Of Measurement)

    Sub New(context As Context, repositories As Repositories, entities As IEnumerable(Of Measurement))
        MyBase.New(context, repositories, entities)
    End Sub

    Public Overrides Function Add(entity As Measurement, ByRef reason As RejectedEntity.RejectedReasons) As Boolean
        Try
            If String.IsNullOrEmpty(entity.WellName) Then
                reason = RejectedEntity.RejectedReasons.WellNameEmpty
                Return False
            ElseIf Not _repositories.Wells.ContainsName(entity.WellName) Then
                reason = RejectedEntity.RejectedReasons.WellNotFound
                Return False
            ElseIf _entities.ContainsKey(entity.Id) Then
                reason = RejectedEntity.RejectedReasons.DuplicatedId
                Return False
            ElseIf entity.FLNADepth > entity.WaterDepth Then
                reason = RejectedEntity.RejectedReasons.FLNADepthGreaterThanWaterDepth
                Return False
            Else
                _entities.Add(entity.Id, entity)
                Dim well = _repositories.Wells.FindName(entity.WellName)
                well.Measurements.Add(entity)
                entity.Well = well
                Return True
            End If
        Catch ex As Exception
            reason = RejectedEntity.RejectedReasons.Unknown
            Return False
        End Try
    End Function
End Class

File: Persistence/CorePersistence/src/Repositories/MeasurementsRepository.cs
using System.Linq;
using Wells.YPFModel;

namespace Wells.CorePersistence.Repositories
{
    public class MeasurementsRepository : RepositoryBase<Measurement>, IBussinessObjectRepository
    {
        public MeasurementsRepository(ApplicationDbContext context) : base(context) { }

        public override bool ContainsName(string name)
        {
            return Exists((c) => c.Name == name);
        }

        public override bool Exists(string id)
        {
            return Exists((c) => c.Id == id);
        }

        public override Measurement Find(string id)
        {
            return Find((c) => c.Id == id);
        }

        public override Measurement FindByName(string name)
        {
            return Find((c) => c.Name == name);
        }

        //protected override RejectedReasons OnAddingOrUpdating(Measurement entity)
        //{
        //    if (string.IsNullOrEmpty(entity.WellName))
        //    {
        //        return RejectedReasons.WellNameEmpty;
        //    }
        //    else if (Exists(entity.Id))
        //    {
        //        return RejectedReasons.DuplicatedId;
        //    }
        //    else if (!RepositoryWrapper.Instance.Wells.ContainsName(entity.WellName))
        //    {
        //        return RejectedReasons.WellNotFound;
        //    }
        //    else if (entity.FLNADepth > entity.WaterDepth)
        //    {
        //        return RejectedReasons.FLNADepthGreaterThanWaterDepth;
        //    }
        //    return RejectedReasons.None;
        //}
    }
}
